flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -2123,6 +2131,92 @@ mlir::Value IntrinsicLibrary::genDprod(mlir::Type resultType,
   return builder.create<mlir::MulFOp>(loc, a, b);
 }
 
+// EOSHIFT
+fir::ExtendedValue
+IntrinsicLibrary::genEoshift(mlir::Type resultType,
+                             llvm::ArrayRef<fir::ExtendedValue> args) {
+  assert(args.size() == 4);
+
+  // Handle required ARRAY argument
+  fir::BoxValue arrayBox = builder.createBox(loc, args[0]);
+  auto array = fir::getBase(arrayBox);
+  auto arrayRank = arrayBox.rank();
+
+  // Create mutable fir.box to be passed to the runtime for the result.
+  auto resultArrayType = builder.getVarLenSeqTy(resultType, arrayRank);
+  auto resultMutableBox =
+      fir::factory::createTempMutableBox(builder, loc, resultArrayType);
+  auto resultIrBox =
+      fir::factory::getMutableIRBox(builder, loc, resultMutableBox);
+
+  // Handle optional BOUNDARY argument
+  mlir::Type arrayEleTy = arrayBox.getEleTy();
+  mlir::Value boundary = nullptr;
+
+  if (isAbsent(args[2])) {
+    if (arrayEleTy.isa<mlir::IntegerType>() ||
+        arrayEleTy.isa<mlir::FloatType>() ||
+        arrayEleTy.isa<fir::LogicalType>()) {
+      // fill in with appropriate number of 0 or .false.
+      auto zero = builder.createIntegerConstant(loc, builder.getIndexType(), 0);
+      auto temp = builder.createTemporary(loc, arrayEleTy);
+      auto cast = builder.createConvert(loc, arrayEleTy, zero);
+      builder.create<fir::StoreOp>(loc, cast, temp);
+      boundary = builder.createBox(loc, temp);
+    } else if (arrayEleTy.isa<fir::CharacterType>()) {
+      // fill in with appropriate number of blanks
+      fir::factory::CharacterExprHelper helper{builder, loc};
+      fir::CharacterType charTy = helper.getCharacterType(arrayEleTy);
+      fir::CharBoxValue temp =
+          helper.createCharacterTemp(arrayEleTy, charTy.getLen());
+      auto zero = builder.createIntegerConstant(loc, builder.getIndexType(), 0);
+      auto len = builder.createIntegerConstant(
+          loc, builder.getCharacterLengthType(), charTy.getLen());
+
+      helper.createPadding(temp, zero, len);
+
+      boundary = builder.createBox(loc, temp);
```

Comment:
I believe the runtime takes care of creating the default boundary itself if it is absent. You only need to pass an absent descriptor argument to the runtime (nullptr). This can be created with:

`boundary = buildear.create<fir::AsentOp>(loc, fir::BoxType::get(builder.getNoneType()))`;

(I am setting the box type is set to None because we do not care about the exact types of absent box).

Reply:
Thank you!  I saw that the runtime handled the default boundary but couldn't figure out how to trigger it to do so, so I codegened everything instead.  This way is much simpler.